hazelcast/src/main/java/com/hazelcast/query/impl/AndResultSet.java

```diff
@@ -161,6 +150,14 @@ public void remove() {
 
     @Override
     public int size() {
-        return setSmallest.size();
+        if (size == -1) {
```

Comment:
this could be problematic performance-wise. as it's used during evaluation. I see you have some sort of caching, but I am not sure about possible consequences. what's the worst (performance) case you can imagine here?